src/pgtest/src/lib.rs

```diff
@@ -205,7 +205,7 @@ impl PgConn {
         &mut self,
         until: Vec<&str>,
         err_field_typs: Vec<char>,
-        ignore: HashSet<String>,
+        ignore: BTreeSet<String>,
```

Comment:
The linked issue describes motivation for removing the iteration of hashmaps, not necessarily their entire usage. This hashmap, for example, only executes `contains` and `insert`, which is not the problem described. I'm not convinced that a wholesale removal of hashmap is the correct solution.

Are there compelling reasons to outright disallow hashmap entirely?

Reply:
Disallowing the use of `std::collections::HashMap` is the only reliable way (I know of) to disallow iterating over a hashmap. In particular, we cannot blacklist the individual iteration methods in clippy.toml, because that way we couldn't cover trait methods like `IntoIterator::into_iter`. Even if we could, we'd need to keep the list of disallowed methods updated as stdlib updates add new ones, which doesn't seem ideal.

The intention is not to disallow the use of hash maps (the data structure) though. As you say, it is valid to use one as a lookup table, without iteration. For this purpose, we provide `mz_ore::collections::HashMap`. It wraps `std:: collections::HashMap` and exports only methods that don't expose unstable iteration order.

While converting the codebase, I defaulted to replacing `HashMap`s with `BTreeMap`s, because that almost always works, as opposed to defaulting to `mz_ore:: collections::HashMap`, which only works in a small subset of instances.

There are two reasons not to use `BTreeMap`:
 * The key type doesn't and cannot implement `Ord`. `Antichain` is an example of such a type.
 * A hash map provides better performance than a BTree map.

The second case is difficult to identify without profiling. As a coarse heuristic one could probably say that a BTree map is faster for small numbers of keys and a hashmap is faster for large numbers of keys. But that really depends on access patterns and the cost of computing a value's hash vs computing it's order.

I think what we can agree to is that on hot paths where we are already using a hashmap, we should not switch to a BTree map without profiling. Even if that means we have to allow unstable iteration.